### PR TITLE
7.0 - Add new modul that calculate leav days according to country and state

### PR DESCRIPTION
**Pull request by OSguard**
...blic holidays and 5 or 6 working days
- [x] need https://github.com/OCA/hr/pull/12 merge first

### CODE REVIEW

**Comment by dreispt** [comment]
Sorry for the pain, but calculation intensive modules should really have some tests.

**Comment by OSguard** [comment]
this would be indeed glade. I m not sure how many time we can spend on writing test that in the end are really meaning full.

**Comment by OSguard** [comment]
this module update a on_change method to make the live more easy for the employee, he needs always check the result. So if the code is part of a automated calculation i agree to write test.
I didn't want to spend more time on this as need to have a clean module.
@dreispt what do you think?

**Comment by dreispt** [comment]
I wont push if you insist you don't want to do it at this moment.
But come on, give it a try! :v:
Here is a example to get you started: https://github.com/OCA/contract/blob/7.0/project_sla/test/project_sla.yml

**Comment by bwrsandman** [comment]
Test fails (#13 ), but unrelated

**Comment by bwrsandman** [comment]
Could you rebase the commit so to pass the tests from #23

**Comment by OSguard** [comment]
Now we fix everything and fix our merge mistakes

**Comment by bwrsandman** [comment]
:+1:

**Comment by pedrobaeza** [comment]
Thanks for the module, Markus. Seems interesting, but as you have introduced the concept of working days, we should go deeper, indicating not only the working days (5 or 6), but what are these days (from monday to saturday, for example), because there are some kind of businesses that the rest day are different from the usual. For example, restaurants, that usually uses monday as non working day.

Another possible issue is strange working calendars, that includes intesive work of 12 hours distributed across only 3 days.

Generalizing to first week working day (from monday to saturday), and last week working day (from tuesday to sunday), we cover each possible case, and the difference between both gives us the number of working days.

Going even deeper, we can extract the getter of these first and last week working days, implemented in the base the way I have said, but allowing to be inherited, so that other modules can calculate this data using algorithms  like rotating schedule working days (for 24h services duty rosters and so on).

What do you think? It's too much to acomplish?

P.S.: Sorry for the long disertation.

File: hr_holidays_weekend/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{
+    "name": "Hr holidays weekend module",
```
Comment (ghost) [inline hr_holidays_weekend/__openerp__.py]: I would change the title for Holidays Weekend.

**Comment by ghost** [comment]
In regard to Pedro's comment, you could use the working schedule in the employee contract to deduce what is the weekend of the employee. Exemple, the employee is scheduled from thuesday to saturday, you can deduce that his weekend is sunday and monday.

**Comment by ntodorova** [comment]
Hi,
we find the idea of Mr. Baeza great! We will start working on a new module that adds the functionality for creating a work week for every employee and connect that functionality to the hr_holidays_weekend module.  :)

**Comment by ghost** [comment]
@ntodorova This functionality is already implemented in module hr_contract. The field is called working_hours.

**Comment by OSguard** [comment]
@dufresnedavid thanks for the hint, we look inside this and will discuss if we simply can/want use this. One think to check is, if using this we need the hole contract management working.

**Comment by dreispt** [comment]
Unless I'm missing something, this is OK to merge.
Can someone else confirm?

**Comment by OSguard** [comment]
@dreispt we are refactor it to use hr_contract as @dufresnedavid mansion, so mark as "work in progress"

**Comment by dreispt** [comment]
OK, updated.

**Comment by OSguard** [comment]
runbot fails with "You try to install module 'hr_payroll_register_report' that depends on module 'report_aeroo'.\nBut the latter module is not available in your system."
How we can fix this or is at just on testing?

File: hr_holidays_weekend/hr_employee.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import fields, orm
+
+
+class hr_employee(orm.Model):
```
Comment (OSguard) [inline hr_holidays_weekend/hr_employee.py]: now we need to remove this, as we switch to hr_contract

File: hr_holidays_weekend/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{
+    "name": "Holidays Weekend",
+    "version": '0.1',
```
Comment (OSguard) [inline hr_holidays_weekend/__openerp__.py]: change to version 1.0

File: hr_holidays_weekend/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        "hr_contract",
+    ],
+    'author': 'initOS GmbH & Co. KG',
+    "category": "",
+    'license': 'AGPL-3',
+    "description": """
+    Excludes weekend days from calculation of holidays
+
+    The module extend hr_employee by adding option for configure
+    the count of working days (5 or 6).
```
Comment (OSguard) [inline hr_holidays_weekend/__openerp__.py]: update description to hr_contract

**Comment by OSguard** [comment]
@dreispt "work in progress" => "needs review" ;)

**Comment by tremlin** [comment]
I tested the code and I experienced some right problems in `_get_number_of_days`. The onchange throws access errors for `hr.contract` and `resource.calendar.attendance` if a user creating a new leave request has not manager rights.

**Comment by OSguard** [comment]
i can confirm, that @ntodorova is a employee to initOS and so cover to our company CLA

**Comment by gurneyalex** [comment]
@OSguard @ntodorova OCA database updated. Thanks for your contribution.

**Comment by mdietrichc2c** [comment]
I want to approve this PR, but there are a lot of `browse()` done without any context... Is it by design or are they just totally unneeded, including eventual localization of dates?

**Comment by mdietrichc2c** [comment]
Okay, after checking all possible issues, I don't see how the lack of context could create an issue; so, LGTM :+1:

File: hr_holidays_weekend/hr_holidays.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                round(math.floor(diff_day))
+        else:
+            result['value']['number_of_days_temp'] = 0
+
+        return result
+
+    def _get_number_of_days(self, cr, uid, date_from, date_to,
+                            tz, tz_offset, employee_id):
+
+        """Returns a float equals to the timedelta between
```
Comment (ghost) [inline hr_holidays_weekend/hr_holidays.py:143]: Use infinitive verbs: Returns -> Return

File: hr_holidays_weekend/hr_holidays.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            result['value']['date_to'] = str(date_to_with_delta)
+
+        # new timezone-shift
+        user_obj = self.pool.get("res.users").browse(cr, uid, uid)
+
+        # Compute and update the number of days
+        if (date_to and date_from) and (date_from <= date_to):
+            diff_day = self._get_number_of_days(cr, uid, date_from, date_to,
+                                                user_obj.tz,
+                                                user_obj.tz_offset,
```
Comment (ghost) [inline hr_holidays_weekend/hr_holidays.py:102]: You can use the context to retreive the time zone.
tz_offset is not used in your function _get_number_of_days, you may remove it.

File: hr_holidays_weekend/hr_holidays.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+        employee = self.pool.get('hr.employee').browse(cr, uid, employee_id)
+
+        hr_holiday_public_obj = self.pool.get('hr.holidays.public')
+        if employee.address_id.country_id is None:
+            country_filter = [('country_id', '=', False)]
+        else:
+            country_filter = ['|', ('country_id', '=',
+                                    employee.address_id.country_id.id),
+                                   ('country_id', '=', False)]
```
Comment (ghost) [inline hr_holidays_weekend/hr_holidays.py:48]: You can simplify this expression a lot. Also, what happends if the employee has no address ?

address = employee.address_id
country = address and address.country_id
country_filter = [('country_id', 'in', [country and country.id, False])]

File: hr_holidays_weekend/hr_holidays.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+            states_filter = [('holidays_id', 'in',
+                              hr_holiday_public),
+                             ('state_ids', '=', False)]
+        else:
+            states_filter = [('holidays_id', 'in',
+                              hr_holiday_public),
+                             '|', ('state_ids.id', '=',
+                                   employee.address_id.state_id.id),
+                             ('state_ids', '=', False)]
+
```
Comment (ghost) [inline hr_holidays_weekend/hr_holidays.py:62]: Same thing here

**Comment by OSguard** [comment]
@ntodorova need to rebase branch

**Comment by OSguard** [comment]
:+1:

**Comment by pedrobaeza** [comment]
@codingforfun, are you going to continue this work?

**Comment by pedrobaeza** [comment]
No answer in a while, so closing. Feel free to reopen if still valid.

**Pull request closed without merging**